### PR TITLE
fix(cbor): guard size arithmetic and zero-value access

### LINKED ISSUE #2094: cbor: guard size arithmetic and zero-value access
CBOR helpers do not handle several boundary and zero-value cases consistently:

- Array header size is understated for lengths at or above the 32-bit boundary.
- Stream position advancement can overflow before its bounds check.
- Methods on an uninitialized lazy value can dereference missing storage.

Acceptance criteria:

- Check size and position arithmetic before conversion or addition.
- Return normal errors for unsupported lengths and advances.
- Define safe zero-value behavior for every exported lazy-value method.
- Add 32-bit, 64-bit, overflow, and zero-value tests.

### PR DESCRIPTION
**Pull request by cryptodj413**
Fixes #2094

## Changes

- `StreamDecoder.Advance`: check for integer overflow before completing
  the position addition, so a huge `n` returns a normal error instead of
  wrapping `newPos` negative and bypassing the existing bounds check.
- `ArrayHeaderSize`: return the correct 9-byte header size for lengths at
  or above the 32-bit boundary (previously understated as 5 bytes for any
  length ≥ 65536, including lengths requiring the 8-byte form).
- `LazyValue`: guard `MarshalCBOR`, `Decode`, `Value`, and `Cbor` against a
  nil internal `*Value` on a zero-value `LazyValue`, matching the lazy-init
  pattern already used in `UnmarshalCBOR` and `MarshalJSON`.

## Testing

- Added 32-bit/64-bit boundary and overflow cases to `TestArrayHeaderSize`
  and `TestStreamDecoderAdvance`.
- Added `TestLazyValueZeroValueMethods` covering all four previously
  unguarded methods on a zero-value `LazyValue`.
- Confirmed each new test panics/fails against the pre-fix code and passes
  with the fix.
- `make test` (root module + all example modules), `go vet`, `golangci-lint`,
  and `nilaway` all pass clean.

---
## Summary by cubic
Fixes three cbor edge cases: `StreamDecoder.Advance` no longer wraps on huge advances, `ArrayHeaderSize` returns the correct 9-byte header for lengths ≥2^32, and all `LazyValue` methods are safe on zero values.

**Changes**
- `StreamDecoder.Advance` now checks the bounds before adding, so a huge `n` returns a normal error instead of wrapping negative and bypassing the check.
- `ArrayHeaderSize` returns 9 for lengths at or above the 32-bit boundary instead of understating as 5.
- `LazyValue` `MarshalCBOR`, `Decode`, `Value`, and `Cbor` guard against a nil internal value, matching the existing lazy-init pattern.
- Added boundary, overflow, and zero-value tests covering all three fixes.

<sup>Written for commit e2bbe8d7d0f9a552ddc59aa82e0ec88b9acfadfc. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/blinklabs-io/gouroboros/pull/2196?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevented stream decoding from accepting invalid advances caused by integer overflow.
  * Corrected array header sizing for lengths above the 32-bit range.
  * Improved safety when using an uninitialized lazy value, avoiding panics and returning appropriate results or errors.

* **Tests**
  * Added coverage for overflow handling, large array lengths, and zero-value lazy value behavior.

### CODE REVIEW

File: cbor/decode.go [hunk 1]
```diff
@@ -395,7 +395,13 @@ func (d *StreamDecoder) Advance(n int) error {
 	if n < 0 {
 		return errors.New("cannot advance by negative amount")
 	}
-	newPos := d.consumed + d.dec.NumBytesRead() + n
+	curPos := d.consumed + d.dec.NumBytesRead()
+	// Check for overflow before completing the addition; a huge n could
+	// otherwise wrap newPos negative and bypass the bounds check below
+	if n > len(d.data)-curPos {
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply cbor/decode.go:401]: Confirmed — after the pre-addition overflow check, `newPos > len(d.data)` can never be true. Removed the redundant second check in e2bbe8d.

File: cbor/stream_decoder_test.go [hunk 2]
```diff
@@ -452,6 +461,12 @@ func TestArrayHeaderSize(t *testing.T) {
 		{256, 3},
 		{65535, 3},
 		{65536, 5},
+		{math.MaxUint32 - 1, 5},
+		{math.MaxUint32, 5},
+		// At and above the 32-bit boundary, CBOR requires the 8-byte
+		// length form (9-byte header), not the 4-byte form (5-byte header).
+		{math.MaxUint32 + 1, 9},
+		{math.MaxInt, 9},
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply cbor/stream_decoder_test.go]: Confirmed — `math.MaxUint32`/`math.MaxUint32+1` are untyped constants that overflow a 32-bit `int`, breaking compilation on GOARCH=386/arm (verified with `GOOS=linux GOARCH=386 go vet ./cbor/...`). Fixed in e2bbe8d: the boundary cases now use explicit `int64` values converted at runtime, and the whole 64-bit-only subtest is skipped via `bits.UintSize < 64`. Re-verified 386/arm vet passes.

File: cbor/value.go [hunk 3]
```diff
@@ -389,14 +392,23 @@ func (l *LazyValue) MarshalJSON() ([]byte, error) {
 }
 
 func (l *LazyValue) Decode() (any, error) {
+	if l.value == nil {
+		l.value = &Value{}
+	}
 	err := l.value.UnmarshalCBOR([]byte(l.value.cborData))
 	return l.Value(), err
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply cbor/value.go:399]: Not reproducible against current code — `LazyValue.Decode()` on a zero-value instance calls `Value.UnmarshalCBOR([]byte(""))`, which already returns `errors.New("empty CBOR data")` (see the len(data)==0 check at the top of UnmarshalCBOR), not `(nil, nil)`. Verified with a manual repro and covered by TestLazyValueZeroValueMethods/Decode, which asserts require.Error(t, err). No change needed.

**Comment by cryptodj413** [comment]
@coderabbitai review

**Review by arepala-uml — APPROVED** [review]
Verified against current head (e2bbe8d), isolated worktree build/test:

- `Advance` overflow fix confirmed: `Advance(math.MaxInt)` now errors instead of wrapping position negative.
- `ArrayHeaderSize` now returns 9 bytes for lengths >= 2^32 (previously understated as 5).
- `LazyValue` zero-value methods (`MarshalCBOR`, `Decode`, `Value`, `Cbor`) no longer nil-deref.
- `go build ./...`, `go test -race ./cbor/...`, full-repo `go test ./...`, and `golangci-lint run ./cbor/...` all pass clean.
- All Cubic/CodeRabbit findings on this PR are resolved or rebutted in-thread and confirmed independently; none remain open.

Non-blocking follow-up (out of scope for this PR): `ledger/common/common.go`'s private `cborArrayHeaderSize` duplicates the same pre-fix header-size bug (caps at 5 bytes for lengths >= 65536). Practically unreachable given Cardano block-size bounds, but same defect class — worth a separate issue.

No merge blockers.
